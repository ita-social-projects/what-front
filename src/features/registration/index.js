export { Registration } from './registration.js';