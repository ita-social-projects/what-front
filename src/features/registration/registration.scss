.wrapper {
  position: absolute;
  width: 100%;
  min-height: 100%;  
  background-color: #020520;  
}

.form {
  background-color: #fff;
  margin-top: 5vh;
  border-radius: 3px;
}