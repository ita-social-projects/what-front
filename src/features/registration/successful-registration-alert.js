import React from 'react';
import { Modal } from 'react-bootstrap';

import classNames from 'classnames';
import styles from './registration.scss';

export const SuccessfulRegistrationAlert = ({toShow, onClose, onSubmit}) => {
  return (
    <Modal
      show={toShow}
      onHide={onClose}
      backdrop="static"
      keyboard={false}
    >
      <Modal.Header>
        <h2>Congratulations</h2>
      </Modal.Header>
      <Modal.Body className={styles["modal-body"]}>
        You have successfully registered.<br/>
        Please, wait until your account is approved and your role is assigned.
      </Modal.Body>
      <Modal.Footer className="d-flex justify-content-center">
        <button
          className={classNames("btn btn-info w-75", styles["modal-btn"])}
          type="submit" onClick={onSubmit}
        >Ok</button>
      </Modal.Footer>
    </Modal>
  );
};