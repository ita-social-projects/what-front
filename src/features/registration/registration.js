import React, { useState, useEffect } from 'react';
import { useActions, paths } from '@/shared/index.js';
import { useSelector, shallowEqual } from 'react-redux';
import { clearRegistration, registration, registrationSelector } from '../../models/index.js';
import { Link, useHistory } from 'react-router-dom';

import { Formik, Form, Field } from 'formik';
import {  validatePassword, validateConfirmPassword, formValidate } from '../validation/validation-helpers.js';
import { Button, WithLoading } from '../../components/index.js';
import { SuccessfulRegistrationAlert } from './successful-registration-alert.js';

import classNames from 'classnames';
import styles from './registration.scss';

export const Registration = () => {
  const history = useHistory();
  const [toShowModal, setShowModal] = useState(false);

  const handleShowModal = () => setShowModal(true);
  const handleCloseModal = () => setShowModal(false);

  const {isLoading, isLoaded, error} = useSelector(registrationSelector, shallowEqual);

  const signUp = useActions(registration);
  const clearLoaded = useActions(clearRegistration);

  const handleSubmitModal = () => {
    handleCloseModal();
    clearLoaded();
    history.push(paths.AUTH);
  };

  const onSubmit = (values) => {
    signUp(values);
  };

  useEffect(() => {
    if (isLoaded && !error) {
      handleShowModal()
    }
  }, [isLoaded, error]);

  return (
    <div className={styles.wrapper}>
      <div className="container">
        <div className="row justify-content-center">
          <div className="col-6 col-md-6 col-sm-12">
            <div className={styles.form}>
            <Formik
              initialValues={{
                firstName: '',
                lastName: '',
                email: '',
                password: '',
                confirmPassword: '',
              }}
              onSubmit={onSubmit}
              validationSchema={formValidate}
            >
              {({ values, errors, touched }) => (
                <Form className="p-3">
                  <h3 className="text-center">Sign up to WHAT</h3>
                  <hr />
                  <div className="form-group">
                    <label htmlFor="firstName">First Name</label>
                    <Field
                      name="firstName"
                      className={classNames('form-control', {'border-danger': errors.firstName && touched.firstName})}
                      id="firstName"
                      placeholder="First name"
                    />
                    <p className="text-danger mb-0">{touched.firstName && errors.firstName}</p>
                  </div>
                  <div className="form-group">
                    <label htmlFor="lastName">Last Name</label>
                    <Field
                      name="lastName"
                      className={classNames('form-control', {'border-danger': errors.lastName && touched.lastName})}
                      id="lastName"
                      placeholder="Last name"
                    />
                    <p className="text-danger mb-0">{touched.lastName && errors.lastName}</p>
                  </div>
                  <div className="form-group">
                    <label htmlFor="email">Email address</label>
                    <Field
                      type="email"
                      name="email"
                      className={classNames('form-control', {'border-danger': errors.email && touched.email})}
                      id="email"
                      placeholder="Email"
                    />
                    <p className="text-danger">{touched.email && errors.email}</p>
                  </div>
                  <div className="form-group">
                    <label htmlFor="password">Password</label>
                    <Field
                      type="password"
                      name="password"
                      validate={validatePassword}
                      className={classNames('form-control', {'border-danger': errors.password && touched.password})}
                      id="password"
                      placeholder="Password"
                      autoComplete="on"
                    />
                    <p className="text-danger mb-0">{touched.password && errors.password}</p>
                  </div>
                  <div className="form-group">
                    <label htmlFor="confirm-password">Confirm Password</label>
                    <Field
                      type="password"
                      name="confirmPassword"
                      validate={(value) => validateConfirmPassword(values.password, value)}
                      className={classNames('form-control', {'border-danger': errors.confirmPassword && touched.confirmPassword})}
                      id="confirm-password"
                      placeholder="Confirm password"
                      autoComplete="on"
                    />
                    <p className="text-danger mb-0">{touched.confirmPassword && errors.confirmPassword}</p>
                  </div>
                  <div className="d-flex justify-content-center">
                      <WithLoading isLoading={isLoading}>
                        <Button type="submit" className="btn btn-block btn-warning">Sign up</Button>
                      </WithLoading>
                    </div>
                  <div className="text-center mt-3">
                    <p>Already have an account? <Link to={paths.AUTH}>Log in</Link></p>
                  </div>
                  <SuccessfulRegistrationAlert 
                    toShow={toShowModal}
                    onClose={handleCloseModal}
                    onSubmit={handleSubmitModal}
                  />
                </Form>
              )}
            </Formik>
            </div>
          </div>
        </div> 
      </div>
    </div>
  );
};