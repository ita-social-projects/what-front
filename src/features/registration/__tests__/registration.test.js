import React from 'react';
import { render, waitFor, fireEvent } from '@testing-library/react';
import { useSelector } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';

import { useActions } from '@/shared/index.js';
import { registration } from '@models/index.js';
import { Registration } from '../registration.js';

jest.mock('react-redux', () => ({ useSelector: jest.fn().mockReturnValue({
  isLoading: false,
  error: '',
  isLoaded: true,
}) }));
jest.mock('@/shared/hooks', () => ({ useActions: jest.fn().mockReturnValue(jest.fn()) }));

describe('Registration', () => {
  let formValues;
  beforeEach(() => {
    formValues = {
      firstName: 'John',
      lastName: 'Smith',
      email: 'testemail@gmail.com',
      password: 'Testpassw!ord123',
      confirmPassword: 'Testpassw!ord123',
    };
  });

  it('should render component', () => {
    const { container } = render(<Router><Registration /></Router>);
    const formContainer = container.getElementsByClassName('container');

    expect(formContainer).toMatchSnapshot();
  });

  it('should validate the form', async () => {
    const {
      getByText,
      getAllByText,
    } = render(<Router><Registration /></Router>);

    const FORM_FIELDS_AMOUNT = 5;
    const submitButton = getByText('Sign up');

    await waitFor(() => {
      fireEvent.click(submitButton);
    });

    const errors = getAllByText('This field is required');

    expect(errors).toHaveLength(FORM_FIELDS_AMOUNT);
    expect(errors).toMatchSnapshot();
  });

  it('should handle error receiving', () => {
    useSelector.mockReturnValue({
      isLoading: false,
      error: 'Something went wrong',
      isLoaded: true,
    });

    const { getByText } = render(<Router><Registration /></Router>);

    const error = getByText('Something went wrong');
    expect(error).toBeInTheDocument();
  });

  it('should submit the form with correct schema', async () => {
    const { getByPlaceholderText, getByText } = render(<Router><Registration /></Router>);

    const firstName = getByPlaceholderText('First name');
    const lastName = getByPlaceholderText('Last name');
    const email = getByPlaceholderText('Email');
    const password = getByPlaceholderText('Password');
    const confirmPassword = getByPlaceholderText('Confirm password');
    const submitButton = getByText('Sign up');

    const signUp = useActions(registration);

    fireEvent.change(firstName, {
      target: {
        value: formValues.firstName,
      },
    });

    fireEvent.change(lastName, {
      target: {
        value: formValues.lastName,
      },
    });

    fireEvent.change(email, {
      target: {
        value: formValues.email,
      },
    });

    fireEvent.change(password, {
      target: {
        value: formValues.password,
      },
    });

    fireEvent.change(confirmPassword, {
      target: {
        value: formValues.confirmPassword,
      },
    });

    await waitFor(() => {
      fireEvent.click(submitButton);
    });

    signUp({
      firstName: firstName.value,
      lastName: lastName.value,
      email: email.value,
      password: password.value,
      confirmPassword: confirmPassword.value,
    });

    expect(signUp).toHaveBeenCalledWith(formValues);
  });
});