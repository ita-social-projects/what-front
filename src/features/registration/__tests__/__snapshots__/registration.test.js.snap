// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Registration should render component 1`] = `
HTMLCollection [
  <div
    class="container"
  >
    <div
      class="row justify-content-center"
    >
      <div
        class="col-xl-6 col-md-8 col-sm-10 col-12"
      >
        <div>
          <form
            action="#"
            class="p-3"
            novalidate=""
          >
            <h3
              class="text-center"
            >
              Sign up to WHAT
            </h3>
            <hr />
            <div
              class="form-group"
            >
              <label
                for="firstName"
              >
                First Name
              </label>
              <input
                class="form-control"
                id="firstName"
                name="firstName"
                placeholder="First name"
                value=""
              />
              <p
                class="text-danger mb-0"
              />
            </div>
            <div
              class="form-group"
            >
              <label
                for="lastName"
              >
                Last Name
              </label>
              <input
                class="form-control"
                id="lastName"
                name="lastName"
                placeholder="Last name"
                value=""
              />
              <p
                class="text-danger mb-0"
              />
            </div>
            <div
              class="form-group"
            >
              <label
                for="email"
              >
                Email address
              </label>
              <input
                class="form-control"
                id="email"
                name="email"
                placeholder="Email"
                type="email"
                value=""
              />
              <p
                class="text-danger"
              />
            </div>
            <div
              class="form-group"
            >
              <label
                for="password"
              >
                Password
              </label>
              <input
                autocomplete="on"
                class="form-control"
                id="password"
                name="password"
                placeholder="Password"
                type="password"
                value=""
              />
              <p
                class="text-danger mb-0"
              />
            </div>
            <div
              class="form-group"
            >
              <label
                for="confirm-password"
              >
                Confirm Password
              </label>
              <input
                autocomplete="on"
                class="form-control"
                id="confirm-password"
                name="confirmPassword"
                placeholder="Confirm password"
                type="password"
                value=""
              />
              <p
                class="text-danger mb-0"
              />
            </div>
            <div
              class="d-flex justify-content-center"
            >
              <button
                class="btn undefined btn btn-block btn-info"
                type="submit"
              >
                Sign up
              </button>
            </div>
            <p
              class="text-danger text-center mt-2"
            />
            <div
              class="text-center mt-3"
            >
              <p>
                Already have an account? 
                <a
                  href="/auth"
                >
                  Log in
                </a>
              </p>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`Registration should validate the form 1`] = `
Array [
  <p
    class="text-danger mb-0"
  >
    This field is required
  </p>,
  <p
    class="text-danger mb-0"
  >
    This field is required
  </p>,
  <p
    class="text-danger"
  >
    This field is required
  </p>,
  <p
    class="text-danger mb-0"
  >
    This field is required
  </p>,
  <p
    class="text-danger mb-0"
  >
    This field is required
  </p>,
]
`;
